node/silkworm/downloader/stage_bodies.cpp

```diff
@@ -0,0 +1,205 @@
+       6.0 index = blockNum - min(requested)
+       6.1 if delivered.contains(blockNum) continue
+       6.2 request_made = requests.contains(offset)
+       6.3 if request_made
+          6.3.1 if not timeout continue
+          6.3.2 else delete request_made from requests, increment peer penalties
+       6.4 header = get_from_deliveries_h_b() or get_from_cache() or get_from_canonical_table(blockNum)
+       6.5 add header to deliveries_h
+       6.6 if block in cache/db
+          6.6.1 add block to deliveries_b,
+          6.6.2 to_request <-- false
+          6.6.3 delivered.add(blockNum)
+       6.7 else
+          6.7.1 to_request <-- true
+          6.7.2 newRequest.blockNums.append(blockNum)
+          6.7.3 newRequest.hashes.append(hash)
+          6.7.4 requests.add(blockNum → newRequest)
+          6.7.5 requestMap.add(hash(txes)+hash(uncles) → blockNum)
+
+    GetDeliveries()
+    ---------------
+    1. for body in received_bodies
+       1.1 if requestedMap[hash(txes)+hash(uncles)] == false
+          1.1.1 continue
+       1.2 else
+          1.2.1 clear requestedMap & requests[offset]
+          1.2.2 deliveriesB.add(offset → body)
+          1.2.3 delivered.add(blockNum)
+    2. headers,bodies = lowest_contiguous_sequence(deliveriesH,deliveriesB)
+    3. remove (headers,bodies) from (deliveriesH,deliveriesB)
+    4. returns (headers,bodies)
+
+ */
+BodiesStage::BodiesStage(const Db::ReadWriteAccess& db_access, BlockDownloader& bd)
+    : db_access_{db_access}, block_downloader_{bd} {
+}
+
+BodiesStage::~BodiesStage() {
+    // todo: implement
+}
+
+Stage::Result BodiesStage::forward([[maybe_unused]] bool first_sync) {
+    using std::shared_ptr;
+    using namespace std::chrono_literals;
+    using namespace std::chrono;
+
+    Stage::Result result;
+
+    StopWatch timing; timing.start();
+    log::Info() << "[2/16 Bodies] Start";
+    log::Trace() << "[INFO] BodyDownloader forward operation started";
+
+    try {
+        Db::ReadWriteAccess::Tx tx = db_access_.start_tx();  // start a new tx only if db_access has not an active tx
+        auto headers_stage_height = tx.read_stage_progress(db::stages::kBlockBodiesKey);
+
+        BodyPersistence body_persistence(tx);
+
+        RepeatedMeasure<BlockNum> height_progress(body_persistence.initial_height());
+        log::Info() << "[2/16 Headers] Waiting for bodies... from=" << height_progress.get();
+
+        auto withdraw_command = withdraw_ready_bodies();
+
+        // block processing
+        time_point_t last_update = system_clock::now();
+        while (body_persistence.highest_height() < headers_stage_height && !block_downloader_.is_stopping()) {
+
+            send_body_requests();
+
+            if (!withdraw_command->completed_and_read()) {
+                // renew request
+                withdraw_command = withdraw_ready_bodies();
+            }
+            else if (withdraw_command->result().wait_for(1000ms) == std::future_status::ready) {
```

Comment:
I suggest using a constant instead of a magic number for the wait timeout.

Reply:
done